manager/allocator/network/allocator.go

```diff
@@ -0,0 +1,869 @@
+	//
+	// Therefore, because the operations can fail can be undone (allocation),
+	// and the operations cannot be undone cannot fail (deallocation) then we
+	// must allocate first, and only deallocate once all allocation succeeds.
+	//
+	// The problem arises specifically because there are TWO operations on
+	// service, allocating ports and allocating VIPs, that can both acquire and
+	// release resources. In an naive implementation, one might do something
+	// like this:
+	//
+	//    Allocate ports
+	//    > Allocate new ports
+	//    > Free unused ports
+	//    Allocate VIPs
+	//    > Allocate new VIPs
+	//    > Release unused VIPs
+	//
+	// Thus, if allocation of new vips fails, we will have already freed some
+	// ports and are stuck in a state where we can't go forward and can't go
+	// backward. To get around this, we have to disallow concurrent access to
+	// one or the other of the port or ipam allocators. Because allocating
+	// ports is likely a less common operation, and because the port allocator
+	// has no external dependencies, it is the part that we choose to
+	// single-thread. Further, because the rolling back logic of a port
+	// allocation is rather complicated and likely to be error prone even if we
+	// guarantee that we can reacquire released ports, we use a two-phase model
+	// for allocating ports
+	//
+	// allocate ports first. calling port.Allocate doesn't actually alter the
+	// state of the port allocator, it just returns a Proposal or an error. If
+	// there is an error then port allocation is not possible, and we can stop
+	// here. Otherwise, the returned proposal contains the information which is
+	// needed to alter the state of the port allocator and commit the changes.
+	// If a proposal is returned, then committing the proposal is guaranteed to
+	// succeed.
+	//
+	// When the code in this function is unpacked, it looks (essentially) like
+	// this:
+	//
+	//   Parts that can fail but be undone:
+	//     port.Allocate: Allocate Ports (verifying if allocation _will_ succeed)
+	//     ipam.AllocateVIPs: Allocate new VIPs
+	//   Parts that cannot fail but cannot be undone:
+	//     in the same call to ipam.AllocateVIPs: Deallocate VIPs no longer in use
+	//     proposal.Commit: Commit the port changes (which we know to be valid)
+	//
+	// This complication does not exist in networks or tasks, which are only
+	// one-way, and therefore only acquire resources or release them, never
+	// both at the same time like services, and it does not exist on nodes
+	// because nodes only worry about one resource type (attachments).
+	proposal, err := a.port.Allocate(endpoint, service.Spec.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	// TODO(dperny) this handles the case of spec.Networks, which we should
+	// deprecate before removing this code entirely
+	networks := service.Spec.Task.Networks
+	if len(service.Spec.Task.Networks) == 0 && len(service.Spec.Networks) != 0 {
```

Comment:
Yes, I suppose we should... I'd really like to just totally deprecate that field, actually, though. I think it's been long enough that it can be done safely. A user would have to have a single service running without ANY spec updates for like 2 years to be covered by that case.